models/src/node/mod.rs

```diff
@@ -148,26 +157,50 @@ impl BottlerocketShadow {
             node_status.current_state == self.spec.state
         })
     }
+
+    /// Returns whether or not a node has crashed.
+    pub fn has_crashed(&self) -> bool {
+        self.status.as_ref().map_or(false, |node_status| {
+            node_status.current_state == BottlerocketShadowState::ErrorReset
+        })
+    }
 }
 
 impl BottlerocketShadowSpec {
     pub fn new(
         state: BottlerocketShadowState,
         state_transition_timestamp: Option<DateTime<Utc>>,
         version: Option<Version>,
+        crash_count: u64,
+        state_transition_failure_timestamp: Option<DateTime<Utc>>,
```

Comment:
I'm thinking it might actually make more sense to include `crash_count` in the `status`, since it has more to do with how the node is currently progressing rather than its desired state.

We may need the nodes themselves to be responsible for updating the crash count, which would help ensure that we don't double-count crashes, but I'm not 100% on this.

I was hoping to draw inspiration from how the `restarts` are stored for Pods in Kubernetes itself, but it seems like they might not be associated directly with `Pod` objects. Still trying to work that out.

Reply:
The reasons that `crash_count` is in the spec are:
1. We have the reset logic that requires the controller to reset the crash_count. And since the time calculate logic is already in the controller, we don't want the agent repeat the time calculate logic for the reset.

2. In a successful update, `crash_count` gets reset during the state transfer MonitoringUpdate -> Idle. On the agent perspective, if the target state is Idle, the agent won't know whether it is from ErrorReset or MonitoringUpdate, which will requires us to track the previous state in the status.